go-controller/pkg/node/udn_isolation.go

```diff
@@ -0,0 +1,459 @@
+				"ip6", "daddr", "@", nftablesUDNPodIPsv6, "accept"),
+		})
+		tx.Add(&knftables.Rule{
+			Chain: nodenft.UDNIsolationChain,
+			Rule: knftables.Concat(
+				"ip6", "daddr", "@", nftablesUDNPodIPsv6, "drop"),
+		})
+	}
+}
+
+func (m *UDNHostIsolationManager) updateKubeletCgroup() error {
+	tx := m.nft.NewTransaction()
+	tx.Flush(&knftables.Chain{
+		Name: nodenft.UDNIsolationChain,
+	})
+	m.addRules(tx)
+
+	err := m.nft.Run(context.TODO(), tx)
+	if err != nil {
+		return fmt.Errorf("could not update nftables rule for management port: %v", err)
+	}
+	return nil
+}
+
+// runKubeletRestartTracker listens to systemd events to re-apply the UDN host isolation rules after kubelet restart.
+// cgroupv2 match doesn't actually match cgroup paths, but rather resolves them to numeric cgroup IDs when such
+// rules are loaded into kernel, and does not automatically update them in any way afterwards.
+// From the patch https://patchwork.ozlabs.org/project/netfilter-devel/patch/1479114761-19534-1-git-send-email-pablo@netfilter.org/#1511797:
+// If the cgroup is gone, the filtering policy would not match anymore. You only have to subscribe to events
+// and perform an incremental updates to tear down the side of the filtering policy that you don't need anymore.
+// If a new cgroup is created, you load the filtering policy for the new cgroup and then add
+// processes to that cgroup. You only have to follow the right sequence to avoid problems.
+func (m *UDNHostIsolationManager) runKubeletRestartTracker(ctx context.Context) (err error) {
+	conn, err := dbus.NewSystemdConnectionContext(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to connect to systemd: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			conn.Close()
+		}
+	}()
+
+	err = conn.Subscribe()
+	if err != nil {
+		return fmt.Errorf("failed to subscribe to systemd events: %w", err)
+	}
+	// interval is important here as we need to catch the restart state, before it is running again
+	events, errChan := conn.SubscribeUnitsCustom(50*time.Millisecond, 0, func(u1, u2 *dbus.UnitStatus) bool { return *u1 != *u2 },
+		func(s string) bool {
+			return s != "kubelet.service"
+		})
+	// run until context is cancelled
+	go func() {
+		waitingForActive := false
+		for {
+			select {
+			case <-ctx.Done():
+				conn.Close()
+				return
+			case event := <-events:
+				for _, status := range event {
+					if status.ActiveState != "active" {
+						waitingForActive = true
+					} else if waitingForActive {
+						klog.Infof("Kubelet was restarted, re-applying UDN host isolation")
+						err = m.updateKubeletCgroup()
+						if err != nil {
+							klog.Errorf("Failed to re-apply UDN host isolation: %v", err)
+						} else {
+							waitingForActive = false
+						}
+					}
+				}
+			case err := <-errChan:
+				klog.Errorf("Systemd listener error: %v", err)
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *UDNHostIsolationManager) podInitialSync() error {
+	pods, err := m.podLister.List(labels.Everything())
```

Comment:
do we need to know about all pods in the cluster?

Reply:
yeah I think so, if you mean compared to per-node filtering.
when the host tries to reach a UDN pod on default network (including pods on other nodes), it will go through mp0. UDN ACLs allow all traffic coming from the management port to allow kubelet traffic, because you can't distinguish host processes in the OVN network. So if you want to avoid node reaching pods on other nodes, you have to drop that traffic on the source node, hence the need to watch for all pods.